ipapython/ipautil.py

```diff
@@ -282,6 +282,37 @@ def write_tmp_file(txt):
 
     return fd
 
+
+@contextmanager
+def concurrent_open(path, mode='r'):
+    """Ensure that complete file is read/written
+
+    Works only for r, rb, w, w+, wb, wb+ modes. Default is r.
+
+    In read mode behaves the same as build-in open. In write mode all data are
+    written to the temporary file first and then moved to target path.
+    This approach ensures that reading will be performed on complete file and
+    writting will result in complete or none file written at all.
+    """
+    if mode in ('w', 'w+', 'wb', 'wb+'):
+        with tempfile.NamedTemporaryFile(
+                mode=mode, prefix=path, delete=False) as temp_file:
+            yield temp_file
+            temp_file.flush()
+            os.fsync(temp_file.fileno())
+            try:
+                os.rename(temp_file.name, path)
+            except EnvironmentError:
```

Comment:
AFAIK EnvironmentError in python 2 is super class for IOError and OSError, in python3 EnvironmentError and IOError are aliases for OSError. So I believe this is correct way of catching any OS/FS-related error in both python 2 and python 3.